<a href="https://colab.research.google.com/github/hansensean123-cell/Sean-Hansen/blob/main/Assignments/assignment_06_data_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# IS 4487 Assignment 6: Data Cleaning with Airbnb Listings

In this assignment, you will:
- Load a raw Airbnb listings dataset
- Identify and resolve missing or inconsistent data
- Decide what data to drop, keep, or clean
- Save a clean dataset to use in Assignment 7

## Why This Matters

Data cleaning is one of the most important steps in any analysis — but it's often the least visible. Airbnb hosts, managers, and policy teams rely on clean data to make decisions. This assignment gives you experience cleaning raw data and justifying your choices so others can understand your process.

<a href="https://colab.research.google.com/github/Stan-Pugsley/is_4487_base/blob/main/Assignments/assignment_06_data_cleaning.ipynb" target="_parent">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>



## Dataset Description

The dataset you'll be using is a **detailed Airbnb listing file**, available from [Inside Airbnb](https://insideairbnb.com/get-the-data/).

Each row represents one property listing. The columns include:

- **Host attributes** (e.g., host ID, host name, host response time)
- **Listing details** (e.g., price, room type, minimum nights, availability)
- **Location data** (e.g., neighborhood, latitude/longitude)
- **Property characteristics** (e.g., number of bedrooms, amenities, accommodates)
- **Calendar/booking variables** (e.g., last review date, number of reviews)

📌 The schema is consistent across cities, so you can expect similar columns regardless of the location you choose.


## 1. Choose a City & Upload Your Dataset

📥 Follow these steps:

1. Go to: [https://insideairbnb.com/get-the-data/](https://insideairbnb.com/get-the-data/)
2. Choose a city you’re interested in.
3. Download the file named: **`listings.csv.gz`** under that city.
4. In your notebook:
   - Open the left sidebar
   - Click the folder icon 📁
   - Click the upload icon ⬆️ and choose your `listings.csv.gz` file
5. Use the file path `/content/listings.csv.gz` when loading your data.
6. Import standard libraries (`pandas`, `numpy`, `seaborn`, `matplotlib`)


In [4]:
# Import necessary libraries 🔧


In [3]:
# Load your uploaded file (path "/content/listings.csv.gz") 🔧
df = pd.read_csv("/listings.csv.gz")
display(df.head())

id                          listing_url       scrape_id last_scraped  \
0  155305  https://www.airbnb.com/rooms/155305  20250617145515   2025-06-17   
1  197263  https://www.airbnb.com/rooms/197263  20250617145515   2025-06-17   
2  209068  https://www.airbnb.com/rooms/209068  20250617145515   2025-06-17   
3  246315  https://www.airbnb.com/rooms/246315  20250617145515   2025-06-17   
4  314540  https://www.airbnb.com/rooms/314540  20250617145515   2025-06-17   

        source                                               name  \
0  city scrape                 Cottage! BonPaul + Sharky's Hostel   
1  city scrape                       Tranquil Room & Private Bath   
2  city scrape                                    Terrace Cottage   
3  city scrape                          Asheville Dreamer's Cabin   
4  city scrape  Asheville Urban Farmhouse Entire Home 4.6 mi t...   

                                         description  \
0  West Asheville Cottage within walking distance...   
1  This is a comfy, peaceful and clean room with ...   
2  Located in one of Asheville's oldest historic ...   
3  Hi there,<br />I am usually here half of each ...   
4  Farmhouse in the city is OPEN! This charming 1...   

                               neighborhood_overview  \
0  We are within easy walk of pubs, breweries, mu...   
1                                                NaN   
2  Our beautiful Grove Park Historic District clo...   
3                                                NaN   
4  City vibes with country appeal. Peaceful neigh...   

                                         picture_url  host_id  ...  \
0  https://a0.muscache.com/pictures/hosting/Hosti...   746673  ...   
1  https://a0.muscache.com/pictures/miso/Hosting-...   961396  ...   
2  https://a0.muscache.com/pictures/1829924/9f3bf...  1029919  ...   
3  https://a0.muscache.com/pictures/5908617/cfe79...  1292070  ...   
4  https://a0.muscache.com/pictures/hosting/Hosti...   381660  ...   

  review_scores_communication review_scores_location review_scores_value  \
0                        4.74                   4.92                4.58   
1                        4.93                   4.86                4.98   
2                        4.98                   4.94                4.80   
3                        4.65                   4.67                4.59   
4                        4.97                   4.91                4.94   

  license instant_bookable calculated_host_listings_count  \
0     NaN                t                              8   
1     NaN                f                              2   
2     NaN                f                              1   
3     NaN                f                              3   
4     NaN                t                              1   

  calculated_host_listings_count_entire_homes  \
0                                           2   
1                                           1   
2                                           1   
3                                           2   
4                                           1   

  calculated_host_listings_count_private_rooms  \
0                                            3   
1                                            1   
2                                            0   
3                                            1   
4                                            0   

  calculated_host_listings_count_shared_rooms reviews_per_month  
0                                           3              2.69  
1                                           0              0.56  
2                                           0              0.40  
3                                           0              0.32  
4                                           0              0.22  

[5 rows x 79 columns]

## 2. Explore Missing Values

Business framing:  

Stakeholders don’t like surprises in the data. Missing values can break dashboards, confuse pricing models, or create blind spots for host managers.

Explore how complete your dataset is:

- Count missing values in each column
- Visualize missingness if helpful (e.g., heatmap, barplot)
- Flag columns that might not be usable due to missing data

### In your markdown:
1. What are the top 3 columns with the most missing values?
2. Which ones are likely to create business issues?
3. Which could be safely ignored or dropped?



In [5]:
# Add code here 🔧


license                         2876
calendar_updated                2876
neighbourhood_group_cleansed    2876
host_neighbourhood              1792
host_about                      1025
neighborhood_overview            952
neighbourhood                    952
host_response_time               442
host_response_rate               442
host_location                    440
price                            340
estimated_revenue_l365d          340
beds                             339
bathrooms                        331
review_scores_cleanliness        263
review_scores_location           263
first_review                     263
review_scores_communication      263
review_scores_rating             263
review_scores_accuracy           263
dtype: int64

### ✍️ Your Response: 🔧
1.License (2876 missing values), calendar updated (2876 missing values), and neighborhood group cleansed (2876 missing values)

2. Missing price information would make it impossible to analyze pricing trends. Missing bed and bathroom amounts would impact search filters affecting booking decisions. Review scores would make it hard to see the listing quality and host performance.

3.License, calendar updated, and neighbourhood group cleaned are all empty, so it wouldn't result in any loss of information for analysis.


## 3. Drop Columns That Aren’t Useful

Business framing:  

Not every column adds value. Analysts often remove columns that are too empty, irrelevant, or repetitive — especially when preparing data for others.

Make a decision:

- Choose 2–4 columns to drop from your dataset
- Document your reasons for each one
- Confirm they're gone with `.head()` or `.info()`

### In your markdown:
1. Which columns did you drop?
2. Why were they not useful from a business perspective?
3. What could go wrong if you left them in?



In [6]:
# Add code here 🔧


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2876 entries, 0 to 2875
Data columns (total 76 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   id                                            2876 non-null   int64  
 1   listing_url                                   2876 non-null   object 
 2   scrape_id                                     2876 non-null   int64  
 3   last_scraped                                  2876 non-null   object 
 4   source                                        2876 non-null   object 
 5   name                                          2876 non-null   object 
 6   description                                   2826 non-null   object 
 7   neighborhood_overview                         1924 non-null   object 
 8   picture_url                                   2876 non-null   object 
 9   host_id                                       2876 non-null   i

### ✍️ Your Response: 🔧
1.I dropped license, calendar updated, and neighbourhood group cleansed since they are all empty.

2.All of those columns were empty with no information so there is no need to have them.

3.It puts more clutter in the dataset which means slower processing and increased memory usage. Potential errors in analysis can arise and make the data more skewed.



## 4. Fill or Fix Values in Key Columns

Business framing:  

Let’s say your manager wants to see a map of listings with prices and review scores. If key fields are blank, the map won’t work. But not all missing values should be filled the same way.

- Choose 2 columns with missing values
- Use a strategy to fill or flag those values
  - (e.g., median, “unknown”, forward-fill, or a placeholder)
- Explain what you did and why

### In your markdown:
1. What two columns did you clean?
2. What method did you use for each, and why?
3. What risks are there in how you filled the data?

In [7]:
# Your code for converting column data types 🔧


/tmp/ipython-input-3637625966.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['price'].fillna(price_median, inplace=True)
/tmp/ipython-input-3637625966.py:8: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try usin

price                   0
review_scores_rating    0
dtype: int64

### ✍️ Your Response: 🔧
1.I cleaned the price and review scores rating columns.

2. I put in the respective median values so the data is less sensitive to outliers when using the mean.

3.Using the mean to fill in missing information can distort the distribution. This is because it makes the distribution narrower and makes it less variable. The median is just an assumed value, so it needs to be taken with a grain of salt.


## 5. Convert and Clean Data Types

Business framing:  

Sometimes columns that look like numbers are actually stored as text — which breaks calculations and slows down analysis. Common examples are price columns with dollar signs or availability stored as strings.

- Identify one column with the wrong data type
- Clean and convert it into a usable format (e.g., from string to number)
- Check your work by summarizing or plotting the cleaned column

### In your markdown:
1. What column did you fix?
2. What cleaning steps did you apply?
3. How does this help prepare the data for later use?


In [8]:
# Clean or adjust your dataset 🔧


float64


count    2876.000000
mean      174.964882
std       209.258133
min        18.000000
25%        95.000000
50%       128.000000
75%       190.000000
max      6846.000000
Name: price, dtype: float64

### ✍️ Your Response: 🔧
1.I cleaned up the price column.

2.I changed the data type from object to numeric then displayed the descriptive stats of it.

3.This helps for later use as I can see the descriptive stats on the price of the airbnb booking and get a better understand of it.

## 6. Remove Duplicate Records

Business framing:  

If a listing appears twice, it could inflate revenue estimates or confuse users. Airbnb needs each listing to be unique and accurate.

- Check for exact duplicate rows
- Also check for duplicate IDs if a unique ID column exists
- Remove duplicates if found and explain your reasoning

### In your markdown:
1. Did you find duplicates?
2. How did you decide what to drop or keep?
3. Why are duplicates risky for Airbnb teams?


In [9]:
# Add code here 🔧


Number of exact duplicate rows: 0
Number of duplicate IDs: 0
No duplicate IDs found.
Number of rows after removing duplicates: 2876


### ✍️ Your Response: 🔧 🔧
1.I didn't find any duplicate rows.

2. I didn't find any duplicates so there wasn't a decision.

3.Duplicates are risky because you don't want the same listing available at one time as that would lead to double-booking and a lot of confusion for hosts and guests.

## 7. Export Cleaned Data

Before wrapping up, export your cleaned Airbnb dataset to a CSV file. You'll need this file for **Assignment 7**, where you'll perform data transformation techniques.

Make sure your data has:
- Cleaned and consistent column values
- Proper data types for each column
- Any unnecessary columns removed

This file should be the version of your dataset that you’d feel confident sharing with a teammate or using for deeper analysis.



```
# Make sure your cleaned DataFrame is named appropriately
# For example, if your cleaned dataset is stored in a DataFrame called 'df_cleaned':

# Export the cleaned DataFrame to a CSV file
df_cleaned.to_csv("cleaned_airbnb_data.csv", index=False)

# Explanation:
# - "cleaned_airbnb_data.csv" is the name of the file that will be saved
# - index=False prevents pandas from writing row numbers into the CSV
# - The file will be saved to your working directory (in Colab, you'll need to download it manually)

```





In [13]:
# export csv here 🔧


## 8. Final Reflection

You’ve just cleaned a real-world Airbnb dataset — the kind of work that happens every day in analyst and data science roles.

Before you move on to data transformation in Assignment 7, take a few moments to reflect on the decisions you made and what you learned.

### In your markdown:
1. What was the most surprising or challenging part of cleaning this dataset?
2. How did you decide which data to drop, fix, or keep?
3. What’s one way a business team (e.g., hosts, pricing analysts, platform ops) might benefit from the cleaned version of this data?
4. If you had more time, what would you explore or clean further?
5. How does this relate to your customized learning outcome you created in canvas?


Write your response clearly in full sentences. No more than a few sentences required per response.


## Submission Instructions
✅ Checklist:
- All code cells run without error
- All markdown responses are complete
- Submit on Canvas as instructed

In [20]:
!jupyter nbconvert --to html "/assignment_06_data_cleaning.ipynb"

[NbConvertApp] Converting notebook /assignment_06_data_cleaning.ipynb to html
[NbConvertApp] Writing 318094 bytes to /assignment_06_data_cleaning.html
